libraries/eosiolib/contracts/eosio/key_value.hpp

```diff
@@ -0,0 +1,896 @@
+               auto copy_size = internal_use_do_not_use::kv_get_data(db_name, 0, (char*)buffer, value_size);
+
+               auto res = memcmp(buffer, tbl_key.data(), copy_size);
+               eosio::check(copy_size == tbl_key.size() && res == 0, "Attempted to update an existing secondary index.");
+
+               if (copy_size > detail::max_stack_buffer_size) {
+                  free(buffer);
+               }
+            } else {
+               auto old_sec_key = table_key(make_prefix(table_name, idx->index_name), idx->get_key(old_value));
+               internal_use_do_not_use::kv_erase(db_name, contract_name.value, old_sec_key.data(), old_sec_key.size());
+               internal_use_do_not_use::kv_set(db_name, contract_name.value, sec_tbl_key.data(), sec_tbl_key.size(), tbl_key.data(), tbl_key.size());
+            }
+         }
+
+      }
+
+      size_t data_size = get_size(value);
+      void* data_buffer = data_size > detail::max_stack_buffer_size ? malloc(data_size) : alloca(data_size);
+
+      serialize(value, data_buffer, data_size);
+
+      internal_use_do_not_use::kv_set(db_name, contract_name.value, tbl_key.data(), tbl_key.size(), (const char*)data_buffer, data_size);
+
+      if (data_size > detail::max_stack_buffer_size) {
+         free(data_buffer);
+      }
+   }
+
+   /**
+    * Removes a value from the table.
+    * @ingroup keyvalue
+    *
+    * @param key - The key of the value to be removed.
+    */
+   void erase(const T& value) {
+      uint32_t value_size;
+
+      auto primary_key = primary_index->get_key(value);
+      auto tbl_key = table_key(make_prefix(table_name, primary_index->index_name), primary_key);
+      auto primary_key_found = internal_use_do_not_use::kv_get(db_name, contract_name.value, tbl_key.data(), tbl_key.size(), value_size);
+
+      if (!primary_key_found) {
+         return;
+      }
+
+      for (const auto& idx : secondary_indices) {
+         auto sec_tbl_key = table_key(make_prefix(table_name, idx->index_name), idx->get_key(value));
+         internal_use_do_not_use::kv_erase(db_name, contract_name.value, sec_tbl_key.data(), sec_tbl_key.size());
+      }
+
+      internal_use_do_not_use::kv_erase(db_name, contract_name.value, tbl_key.data(), tbl_key.size());
+   }
+
+protected:
+   kv_table() = default;
+
+   void setup_indices() {}
+
+   template <typename I, typename... Indices>
+   void setup_indices(I index, Indices... indices) {
+      index->contract_name = contract_name;
+      index->table_name = table_name;
+      index->tbl = this;
+
+      index->setup();
+      secondary_indices.push_back(index);
+      setup_indices(indices...);
+   }
+
+   template <typename PrimaryIndex, typename... SecondaryIndices>
+   void init(eosio::name contract, eosio::name table, eosio::name db, PrimaryIndex prim_index, SecondaryIndices... indices) {
+      validate_types(prim_index, indices...);
```

Comment:
A fold expression is somewhat more friendly than recursion, as it reduces the depth of the template instantiation stack in error messages.
You might also consider ``std::enable_if_t<(std::is_base_of_v<kv_index, std::remove_ptr_t<SecondaryIndices>> && ...)>``

When we get C++20, you'll be able to use concepts for this.

Reply:
Using `enable_if` for both `PrimaryIndex` and `SecondayIndices` gets a bit unwieldy, so I'll probably just stick with a fold expression here.